sources/zendesk/helpers/backloading.py

```diff
@@ -0,0 +1,24 @@
+from typing import Dict, Any
+
+from dlt.common import pendulum
+
+from dlt.common.configuration.specs import BaseConfiguration, configspec
+
+
+@configspec
+class SliceConfig(BaseConfiguration):
+    start: pendulum.DateTime = None
```

Comment:
by this we are creating a new patter how to handle backloading and incremental. if we do this we should update all the other pipelines. currently we have following pattern:
- you can pass `start_date` or `initial_load_past_days` which will be used to initialize `initial_value` of incremental
- if not present incremental loads from the beginning
- the `start_date` is obviously ignored on any next pipeline run - incremental will use the last value to continue
- we didn't have option for `end_date`. that obviously would need to disable incremental altogether

from the user perspecitve the biggest advangtage is that you can jus run pipeline in a loop. the backloading and subsequent incremental runs use the same pipeline code

when I look how current thing works
- you can load a time period (from date to date). incremental is disabled
- or you can load from the beginning with incremental. so there's no way to backload and then use incremental
- or you need to build some custom DAGs where you do backloading and then incremental. but even in that case you need to pass the end time of last backload to the incremental run as initial value

maybe we could combine this patterns
- if we have end date then no incremental
- if we have start data but no end date then we use the start date as initial and incremental is enable
- if no range specified then incremental starts from the beginning (or some other default which needs to be documented)

Reply:
Considering this, probably these should be `incremental` params and we add this logic there.  

I.e. add to `dlt.sources.incremental` both `start_value` and `end_value` arguments.  
The start value would always supersede the state value (`initial_value` behaviour doesn't change)

Usage as e.g.

```python
@source
def zendesk_support(start_date: DateTime=SOME_DEFAULT, end_date: DateTime = None):
    if end_date:
        incremental_args = dict(start_value=start_date, end_value=end_date)
    else::
        incremental_args = dict(initial_value=start_date)
    
    yield ticket_resource(updated_at=dlt.sources.incremental[DateTime](**incremental_args))
```

The filtering in incremental would be pretty much the same I think, couple of differences when start/end values are supplied:

1. Only write `end_value` to state

    Instead of writing the current row value to state, we only write the highest `end_value` to state   
    e.g.  
     ```python
     new_state_value = self.last_value_func([end_value, current_state_value])
     ```
    -> Makes it possible to switch from chunked backloading to incremental. It just starts from end of highest range  
    
2. Stop the generator when end value is reached

    tbd how to do this, something like this is available in sources `add_limit`   
    Check e.g.
    
    ```python
    if row_value == end_value or self.max_value_func([row_value, self.end_value]) != self.end_value:
        # Reached end value, stop
        kill_generator_somehow()
    ```
    
This would mean you can backload, and then stop supplying the start/end args to go back to incremental.  
Or vice versa you can load incrementally new data and then backfill later.  

    
The only iffy part about this I can think of is if you're gonna be loading chunks in parallel. Since it's writing to state and there's no synchronization there. But solvable by e.g. loading the highest chunk first.

Reply:
parallel will not be possible in all cases and that's OK - in airflow, the user would regulate that with https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#config-core-max-active-runs-per-dag